packages/grafana-data/src/transformations/transformers/joinDataFrames.ts

```diff
@@ -0,0 +1,294 @@
+  /**
+   * The input fields
+   */
+  frames: DataFrame[];
+
+  /**
+   * The field to join -- frames that do not have this field will be droppped
+   */
+  joinBy?: FieldMatcher;
+
+  /**
+   * Optionally filter the non-join fields
+   */
+  keep?: FieldMatcher;
+
+  /**
+   * When the result is a single frame, this will to a quick check to see if the values are sorted,
+   * and sort if necessary.  If the first/last values are in order the whole vector is assumed to be
+   * sorted
+   */
+  enforceSort?: boolean;
+
+  /**
+   * @internal -- used when we need to keep a reference to the original frame/field index
+   */
+  keepOriginIndices?: boolean;
+}
+
+/**
+ * This will return a single frame joined by the first matching field.  When a join field is not specified,
+ * the default will use the first time field
+ */
+export function outerJoinDataFrames(options: JoinOptions): DataFrame | undefined {
+  if (!options.frames?.length) {
+    return undefined;
+  }
+  const joinFieldMatcher = options.joinBy ?? pickBestJoinField(options.frames);
+
+  if (options.frames.length === 1) {
+    const frame = options.frames[0];
+    if (options.enforceSort) {
+      const joinIndex = frame.fields.findIndex((f) => joinFieldMatcher(f, frame, options.frames));
+      if (joinIndex >= 0) {
+        if (!isLikelyAscendingVector(frame.fields[joinIndex].values)) {
+          return sortDataFrame(frame, joinIndex);
+        }
+      }
+    }
+    return frame;
+  }
+
+  const nullModes: JoinNullMode[][] = [];
+  const allData: AlignedData[] = [];
+  const originalFields: Field[] = [];
+
+  for (let frameIndex = 0; frameIndex < options.frames.length; frameIndex++) {
+    const frame = options.frames[frameIndex];
+    if (!frame || !frame.fields?.length) {
+      continue; // skip the frame
+    }
+    const nullModesFrame: JoinNullMode[] = [NULL_REMOVE];
+
+    let join: Field | undefined = undefined;
+    let fields: Field[] = [];
+    for (let fieldIndex = 0; fieldIndex < frame.fields.length; fieldIndex++) {
+      const field = frame.fields[fieldIndex];
+      getFieldDisplayName(field, frame, options.frames); // cache displayName in state
+
+      if (!join && joinFieldMatcher(field, frame, options.frames)) {
+        join = field;
+      } else {
+        if (options.keep && !options.keep(field, frame, options.frames)) {
+          continue; // skip field
+        }
+
+        // Support the standard graph span nulls field config
+        nullModesFrame.push(field.config.custom?.spanNulls ? NULL_REMOVE : NULL_EXPAND);
+
+        let labels = field.labels ?? {};
+        if (frame.name) {
+          labels = { ...labels, name: frame.name };
+        }
+
+        fields.push({
+          ...field,
+          labels, // add the name label from frame
+        });
+      }
+
+      if (options.keepOriginIndices) {
+        field.state!.origin = {
+          frameIndex,
+          fieldIndex,
+        };
+      }
+    }
+
+    if (!join) {
```

Comment:
not a huge fan of negating stuff that may or may not be objects, always preferred `join != null`, but can leave as-is.